src/base64.cr

```diff
@@ -244,65 +244,89 @@ module Base64
+    while (size > 0) && (sym = bytes[size - 1]) && (sym == NL || sym == NR || sym == PAD)
       size -= 1
     end
-    endcstr = cstr + size - 4
 
+    # Process combinations of four characters until there aren't any left
+    fin = bytes + size - 4
     while true
-      break if cstr > endcstr
-      while cstr.value == NL || cstr.value == NR
-        cstr += 1
-      end
+      break if bytes > fin
 
-      break if cstr > endcstr
-      a, b, c, d = next_decoded_value, next_decoded_value, next_decoded_value, next_decoded_value
+      # Move the pointer by one byte until there is a valid base64 character
+      while bytes.value == NL || bytes.value == NR
+        bytes += 1
+      end
+      break if bytes > fin
 
-      yield (a << 2 | b >> 4).to_u8!
-      yield (b << 4 | c >> 2).to_u8!
-      yield (c << 6 | d).to_u8!
+      yield_decoded_chunk_bytes(
+        raise_error: true,
+        chunk_pos: bytes - bytes_begin,
+        bytes: {bytes[0], bytes[1], bytes[2], bytes[3]}
+      )
+      bytes += 4
     end
 
-    while (cstr < endcstr + 4) && (cstr.value == NL || cstr.value == NR)
-      cstr += 1
+    # Move the pointer by one byte until there is a valid base64 character or the end of `bytes` was reached
+    while (bytes < fin + 4) && (bytes.value == NL || bytes.value == NR)
+      bytes += 1
     end
 
-    mod = (endcstr - cstr) % 4
-    if mod == 2
-      a, b = next_decoded_value, next_decoded_value
-      yield (a << 2 | b >> 4).to_u8!
-    elsif mod == 3
-      a, b, c = next_decoded_value, next_decoded_value, next_decoded_value
-      yield (a << 2 | b >> 4).to_u8!
-      yield (b << 4 | c >> 2).to_u8!
-    elsif mod != 0
-      raise Error.new("Wrong size")
+    # If the amount of base64 characters is not divisable by 4, the remainder of the previous loop is handled here
+    unread_bytes = (fin - bytes) % 4
+    case unread_bytes
+    when 1
+      raise Base64::Error.new("Wrong size")
+    when 2
+      yield_decoded_chunk_bytes(
+        raise_error: true,
+        chunk_pos: bytes - bytes_begin,
+        bytes: {bytes[0], bytes[1]}
+      )
+    when 3
+      yield_decoded_chunk_bytes(
+        raise_error: true,
+        chunk_pos: bytes - bytes_begin,
+        bytes: {bytes[0], bytes[1], bytes[2]}
+      )
     end
   end
 
-  private macro next_decoded_value
-    sym = cstr.value
-    res = dt[sym]
-    cstr += 1
-    if res < 0
-      raise Error.new("Unexpected byte 0x#{sym.to_s(16)} at #{cstr - start_cstr - 1}")
-    end
-    res
+  # This macro decodes the given chunk of (2-4) base64 characters.
+  # The first argument (raise_error) determines if the macro raises an error when an invalid character is processed.
+  # The second argument (chunk_pos) is only used for the resulting error message.
+  # The resulting bytes are then each yielded to the given block, using the variable `byte`.
+  private macro yield_decoded_chunk_bytes(raise_error, chunk_pos, bytes, &block)
```

Comment:
This is really a minor detail, but I think it's easier to understand this macro if the `bytes` parameter comes first. Because those bytes are what's yielded. Makes sense to put that up front as a positional argument.

The `raise_error` parameter seems unnecessary because all calls pass `true`. The block param is also obsolete.

`bytes` could really be a splat arg. That fits very well:
```suggestion
  private macro yield_decoded_chunk_bytes(*bytes, chunk_pos)
```

Call site looks like this:

```cr
yield_decoded_chunk_bytes(bytes[0], bytes[1], bytes[2], chunk_pos: bytes - bytes_begin)
```